### PR TITLE
supporting cases of multiselect for geneExpression and metaboliteIntensity

### PR DESCRIPTION
**Pull request by airenzp**
## Description

please pull sjpp branch `hierCluster-url-fix` to test hiercluster links, as plot state has been changed #stjude/sjpp#386

This PR allows to select multiple terms in the term type search, and uses it to support geneExpression and metaboliteIntensity cases of multiselect. There is a major refactor to support different term types in the hierCluster and to load top terms for different types. Now hierCluster for metabolites works. Please check
## Checklist

[Check each task](https://github.com/stjude/proteinpaint/wiki/Pull-Request-Checklist) that has been performed or verified to be not applicable.
- [x] Tests: added and/or passed unit and integration tests, or N/A
- [x] Todos: commented or documented, or N/A
- [x] Notable Changes: updated release.txt, prefixed a commit message with "fix:" or "feat:", added to an internal tracking document, or N/A

### CODE REVIEW

File: client/plots/hierCluster.config.js [hunk 1]
```diff
@@ -18,7 +18,7 @@ export async function getPlotConfig(opts = {}, app) {
 		- non-gene genomic stuff that resolves into numeric quantities (cpg meth)
 		- metabolite
 		*/
-		dataType: dtgeneexpression,
+		dataType: opts.termgroups?.[0]?.termType || TermTypes.GENE_EXPRESSION,
```
Comment (xzhou82) [inline client/plots/hierCluster.config.js:21]: code that's launch hiercluster should be required to specify datatype. if missing should abort, and not to default to gene exp

File: client/plots/hierCluster.renderers.js [hunk 2]
```diff
@@ -1,14 +1,14 @@
...
-	switch (this.config.settings.hierCluster.dataType) {
-		case dtgeneexpression:
-			this.dom.header.text('Gene Expression Clustering')
+	switch (this.config.dataType) {
+		case TermTypes.METABOLITE_INTENSITY:
+			this.dom.header.text('Metabolite Intensity Clustering')
 			break
 		default:
-			throw 'unknown hierCluster.dataType to set header'
+			this.dom.header.text('Gene Expression Clustering')
```
Comment (xzhou82) [inline client/plots/hierCluster.renderers.js:11]: should declare all supported data types in switch(), and throw in default

File: client/mass/charts.js [hunk 3]
```diff
@@ -209,6 +208,34 @@ function getChartTypeList(self, state) {
 			}
 		}
 	]
+	if (state.termdbConfig.allowedTermTypes.includes(TermTypes.METABOLITE_INTENSITY)) {
```
Comment (xzhou82) [inline client/mass/charts.js:211]: here u conditionally add button. this deviates from the existing pattern where all possible buttons are created, and later in line 53 main() are changed in visibility. should keep using this pattern due to its merit of easily show/hide buttons on state change. could be that on surv portal on changing cohort (ccss to sjlife), the MRI button will appear as the MRI datatype only exists in that case.

File: client/plots/hierCluster.config.js [hunk 4]
```diff
@@ -161,6 +161,40 @@ export function makeChartBtnMenu(holder, chartsInstance) {
 					return tw
 				})
 			)
+
+			if (tws.length == 1) {
```
Comment (xzhou82) [inline client/plots/hierCluster.config.js:165]: may do away with this makeChartBtnMenu() of hierCluster, and implement the Gene Expression chart button logic in a consistent way as metabolicIntensity as in mass/charts.js

File: client/mass/charts.js [hunk 5]
```diff
@@ -174,7 +173,7 @@ function getChartTypeList(self, state) {
 			clickTo: self.loadChartSpecificMenu
 		},
 		{
-			label: 'Hierarchical Clustering',
+			label: 'Gene Expression',
 			chartType: 'hierCluster',
```
Comment (xzhou82) [inline client/mass/charts.js:177]: see my comment in hierCluster.config.js, not reasonable to still associate `hierCluster` as chart type for this button. should follow logic of metabolite

File: server/shared/common.js [hunk 6]
```diff
@@ -32,6 +32,7 @@ export const dtnloss = 8
 export const dtcloss = 9
 export const dtloh = 10 // to be used in svcnv track
 export const plotColor = '#ce768e'
+export const dtmetaboliteintensity = 11
```
Comment (xzhou82) [inline server/shared/common.js:35]: yes i agree with this design
later can introduce `dtdictionaryterms` and `dtsinglecellgeneexp` and support those in hiercluster

File: client/plots/hierCluster.js [hunk 7]
```diff
@@ -114,7 +115,10 @@ export class HierCluster extends Matrix {
 					values: [
 						{
 							sample: column.name,
-							dt: this.state.config.dataType,
+							dt:
+								this.state.config.dataType == TermTypes.GENE_EXPRESSION
```
Comment (xzhou82) [inline client/plots/hierCluster.js:119]: defaults to one dt is always no good here. can introduce a new mapping from TermTypes to dt
`dt: TermType2Dt(this.state.config.dataType)` 
in which TermType2Dt() throws on unknown type

File: client/mass/charts.js (hunk 3 above)
Comment (airenzp) [reply client/mass/charts.js:211]: the logic for hiding/showing this button cant be solved checking this.state.supportedChartTypes.includes(d.chartType). There is no metaboliteIntensity chartType. This button decides the chartType dynamically, the same happens with the geneExpression button.

Comment (xzhou82) [reply client/mass/charts.js:211]: it should be fixed when computing supportedChartTypes around line 503 of termdb.server.init.js

Comment (airenzp) [reply client/mass/charts.js:211]: do you mean I should create a fake chartType?

File: client/dom/genesetEdit.ts [hunk 8]
```diff
@@ -251,13 +251,18 @@ export function showGenesetEdit(arg: showGenesetEditArg) {
 	}
 
 	function renderGenes() {
+		const orderedGenes = geneList.sort((a, b) => {
```
Comment (xzhou82) [inline client/dom/genesetEdit.ts:254]: when `gene.mutationStat[]` is present, please do not sort alphabetically as genes are already ranked on the server according to stat

also mutationStat is not declared in the type def

Comment (airenzp) [reply client/dom/genesetEdit.ts]: Thanks for pointing this out. Done.

File: client/mass/charts.js [hunk 9]
```diff
@@ -286,9 +323,109 @@ function setRenderers(self) {
 		})
 	}
 
+	self.showGenesetEditUI = async chart => {
```
Comment (xzhou82) [inline client/mass/charts.js:326]: can comment that right now, this ui is hardcoded to produce geneExpression terms. later may generalize it for other gene-based quantification data

File: client/plots/hierCluster.js [hunk 10]
```diff
@@ -89,7 +90,7 @@ export class HierCluster extends Matrix {
 				// for now backend returns {gene:str, data:{}} if there's only 1 eligible gene
 				throw `Cannot do clustering: data is only available for 1 gene (${d.gene}). Try again by adding more genes.`
 			}
-			throw 'Cannot do clustering: invalid server response (lacks .clustering{})'
+			//throw 'Cannot do clustering: invalid server response (lacks .clustering{})'
```
Comment (xzhou82) [inline client/plots/hierCluster.js:93]: should this be reverted or deleted?

Comment (xzhou82) [reply client/plots/hierCluster.js:93]: if so please revert the change

Comment (airenzp) [reply client/plots/hierCluster.js:93]: I meant it can not throw if there is no gene property. It is ok like it is.

**Review by xzhou82 — APPROVED** [review]
looks good. can merge now and we have until Friday to root out bugs

**Pull request merged**